### DISCUSSION TITLE
Get the datanode info by OzoneClient API to check SCMContainerPlacementRackScatter

### DISCUSSION THREAD

**Discussion by Montura** [apache/ozone#6362] [FAQ] [answered]
Hi, community!

I have to read files using `OzoneClient` and then verify, on which datanodes these files are distributed. Is it possible to get info about DNs (and|or about racks) by `OzoneKey`, `OzoneBucket` or `OzoneClient` APIs? 

_In details:_ I'd like to check the scenario with `SCMContainerPlacementRackScatter`. I expect that if I wrote file with `RATIS_THEE` replication factor then each **replica will be saved on different racks**: 
- _ex.: if I have cluster with 3 racks, than each rack will have exactly one replica of a container. And I'd like to make sure that there are no empty racks (without any replicated containers) after writing data._

**Comment by siddhantsangwan** [comment]
How about `OzoneBucket#getFileStatus`? Or using `OzoneBucket#getKey` to get the containers the key's blocks are in, and then finding out which datanodes have these containers?

**Reply by Montura** [reply]
@siddhantsangwan, thanks, I actually took a look at this methods. But I'm not sure for the correct way.

I found two possible approaches:
1. `OzoneBucket#getKey`, and then `OzoneKey#getOzoneKeyLocations`. But then I don't know how to use Ozone API to get find out which datanodes have these containers.

```
    OzoneKeyDetails key = bucket.getKey("key");
    List<OzoneKeyLocation> ozoneKeyLocations = key.getOzoneKeyLocations();
    for (OzoneKeyLocation keyLocation : ozoneKeyLocations) {
      long containerID = keyLocation.getContainerID();
      long localID = keyLocation.getLocalID();
      long length = keyLocation.getLength();
      long offset = keyLocation.getOffset();
      long keyOffset = keyLocation.getKeyOffset();
    }
```

2. `OzoneBucket#getFileStatus` -> `OzoneFileStatus#getKeyInfo` -> `OmKeyInfo#getLatestVersionLocations` -> `OmKeyLocationInfoGroup#getLocationList` -> `OmKeyLocationInfo#getPipeline` -> `Pipeline#getNodes` -> `DatanodeDetails#getNetworkLocation`. 
```
    OzoneClient client = Mockito.mock(OzoneClient.class);
    ObjectStore objectStore = client.getObjectStore();
    OzoneVolume volume = objectStore.getVolume("volume");
    OzoneBucket bucket = volume.getBucket("bucket");
    OzoneFileStatus fileStatus = bucket.getFileStatus("key");
    OmKeyInfo keyInfo = fileStatus.getKeyInfo();
    OmKeyLocationInfoGroup latestVersionLocations = keyInfo.getLatestVersionLocations();
    List<OmKeyLocationInfo> locationList = latestVersionLocations.getLocationList();
    for (OmKeyLocationInfo omKeyLocationInfo : locationList) {
      Pipeline pipeline = omKeyLocationInfo.getPipeline();
      List<DatanodeDetails> nodes = pipeline.getNodes();
      for (DatanodeDetails node : nodes) {
        System.out.println("node = " + node.getNetworkLocation());
      }
    }
```

**Question:** The second way is ok, am I right?

**Reply by siddhantsangwan** [reply, marked as answer]
Yeah looks ok to me. The java doc of `getLocationList` suggests using `getLocationLists` instead:
```
/**
   * Use this expensive method only when absolutely needed!
   * It creates a new list so it is not an O(1) operation.
   * Use getLocationLists() instead.
   * @return a list of OmKeyLocationInfo
   */
```
